File: sdk/src/paginator.ts
import type { Pagination } from './types.js';

export interface PageResult<T> {
  data: T[];
  pagination: Pagination;
}

export type PageFetcher<T, P> = (params: P & { page: number; limit: number }) => Promise<PageResult<T>>;

/**
 * Async generator that automatically fetches subsequent pages.
 *
 * @example
 * for await (const asset of client.assets.listAll({ limit: 100 })) {
 *   console.log(asset.id);
 * }
 */
export async function* paginate<T, P extends { page?: number; limit?: number }>(
  fetcher: PageFetcher<T, P>,
  params: P
): AsyncGenerator<T> {
  let page = params.page ?? 1;
  const limit = params.limit ?? 50;

  while (true) {
    const result = await fetcher({ ...params, page, limit });
    for (const item of result.data) {
      yield item;
    }
    if (page >= result.pagination.pages) break;
    page++;
  }
}

File: sdk/python/servicecycle/paginator.py
"""Auto-paginating iterator helper. Mirrors sdk/src/paginator.ts."""

from __future__ import annotations

from typing import Any, Callable, Dict, Iterator


def paginate(fetcher: Callable[[Dict[str, Any]], Dict[str, Any]], params: Dict[str, Any]) -> Iterator[Any]:
    """Yield items across all pages of a `{data, pagination}` list endpoint.

    Example::

        for asset in client.assets.list_all(limit=100):
            print(asset["id"])
    """
    page = params.get("page") or 1
    limit = params.get("limit") or 50
    while True:
        result = fetcher({**params, "page": page, "limit": limit})
        for item in result["data"]:
            yield item
        pagination = result["pagination"]
        if page >= pagination["pages"]:
            break
        page += 1
